File: src/main/java/com/github/morningwn/t0001_t0500/t0001_t0100/t0061_t0070/T0061.java
package com.github.morningwn.t0001_t0500.t0001_t0100.t0061_t0070;

/**
 * <a href="https://leetcode-cn.com/problems/plus-one/">66. 加一</a>
 *
 * @author moringwn
 * @date 2021/10/21 上午9:50
 */
public class T0061 {

    public int[] plusOne(int[] digits) {

        int carry = 1;
        int tmp = 0;

        for (int i = digits.length - 1; i >= 0; i--) {
            tmp = digits[i];
            digits[i] = (tmp + carry) % 10;
            carry = (tmp + carry) / 10;
        }

        if (carry == 1) {
            int[] ints = new int[digits.length + 1];
            System.arraycopy(digits, 0, ints, 1, digits.length);
            ints[0] = carry;
            return ints;
        } else {
            return digits;
        }

    }
}

File: src/main/c/t0001_t0500/t0001_t0100/t0061_t0070/T0061.c
/**
 * 66. 加一
 * https://leetcode-cn.com/problems/plus-one/
 */
#include <stdio.h>
#include <stdlib.h>

int *plusOne(int *digits, int digitsSize, int *returnSize) {
    int carry = 1;
    int tmp = 0;

    for (int i = digitsSize - 1; i >= 0; i--) {
        tmp = digits[i];
        digits[i] = (tmp + carry) % 10;
        carry = (tmp + carry) / 10;
    }

    if (carry != 0) {
        *returnSize = digitsSize + 1;
        digits = (int *) realloc(digits, *returnSize * sizeof(int));
        digits[0] = 1;
        digits[*returnSize - 1] = 0;

        return digits;
    } else {
        *returnSize = digitsSize;
        return digits;
    }
}

int main() {

    int nums[] = {9, 9, 9, 9};
    int returnSize = 4;

    int *res = plusOne(nums, 4, &returnSize);

    printf("returnSize: %d \n", returnSize);

    for (int i = 0; i < returnSize; ++i) {
        printf("%d, ", res[i]);
    }

    return 0;
}
